spring-cloud-gcp-core/src/main/java/org/springframework/cloud/gcp/core/GcpEnvironment.java

```diff
@@ -35,43 +30,64 @@
 	/**
 	 * Matches Kubernetes instances.
 	 */
-	KUBERNETES_ENGINE(() -> System.getenv("KUBERNETES_SERVICE_HOST") != null),
+	KUBERNETES_ENGINE,
 
 	/**
 	 * Matches App Engine Flexible instances.
 	 */
-	APP_ENGINE_FLEXIBLE(() -> System.getenv("GAE_INSTANCE") != null),
+	APP_ENGINE_FLEXIBLE,
 
 	/**
 	 * Matches App Engine Standard instances.
 	 */
-	APP_ENGINE_STANDARD(() -> ServiceOptions.getAppEngineAppId() != null),
+	APP_ENGINE_STANDARD,
 
 	/**
 	 * Matches instances of Compute Engine that are not also AppEngine Flexible.
 	 */
-	COMPUTE_ENGINE(() -> MetadataConfig.getInstanceId() != null && !APP_ENGINE_FLEXIBLE.matches()),
+	COMPUTE_ENGINE,
 
 	/**
 	 * Matches both, AppEngine Flexible and AppEngine Standard.
 	 */
-	ANY_APP_ENGINE(() -> APP_ENGINE_FLEXIBLE.matches() || APP_ENGINE_STANDARD.matches()),
+	ANY_APP_ENGINE {
+		public boolean matchesSimpleEnvironment(GcpEnvironment env) {
+			validateSimpleEnvironment(env);
+			return APP_ENGINE_FLEXIBLE == env || APP_ENGINE_STANDARD == env;
+		}
+	},
 
 	/**
 	 * Matches both, Compute Engine and Kubernetes Engine.
 	 */
-	GKE_OR_GCE(() ->  KUBERNETES_ENGINE.matches() || COMPUTE_ENGINE.matches());
-
-	private BooleanSupplier matchCondition;
+	GKE_OR_GCE {
+		public boolean matchesSimpleEnvironment(GcpEnvironment env) {
+			validateSimpleEnvironment(env);
+			return COMPUTE_ENGINE == env || KUBERNETES_ENGINE == env;
+		}
+	},
 
-	GcpEnvironment(BooleanSupplier matchCondition) {
-		this.matchCondition = matchCondition;
-	}
+	/**
+	 * Matches nothing; environment cannot be identified.
+	 */
+	UNKNOWN {
+		public boolean matchesSimpleEnvironment(GcpEnvironment env) {
+			return false;
+		}
+	};
 
 	/**
 	 * @return whether the specified environment's presence is detected.
 	 */
-	public boolean matches() {
-		return this.matchCondition.getAsBoolean();
+	public boolean matchesSimpleEnvironment(GcpEnvironment env) {
+		validateSimpleEnvironment(env);
+		return this == env;
+	}
+
+	private static void validateSimpleEnvironment(GcpEnvironment env) {
+		if (env == ANY_APP_ENGINE || env == GKE_OR_GCE) {
+			throw new IllegalArgumentException("Composite environments are not allowed here.");
```

Comment:
I don't think `here` is so informative word.
I guess we need to be more specific to explain end-user what is wrong.

Thanks